### PR TITLE
add sd/ex2c.c to implement 2D HR solvers

### PR DESCRIPTION
**Pull request by donghuix**
Several developments in sd/ex2c.c on the top of swe/ex2b.c to implement 2D Hairsine-Rose equations:
1. Hard code boundary code for outlet
2. Write flow and sediment discharge into output file
3. add new outflow boundary condition: ``SOFT_BOUNDARY`` (i.e., assuming ``hr = hl``, ``ur = ul``, ``vr = vl``)
4. Solver for sediment 
5. Matlab code to visualize validation against tRIBS-FEaST

Run the ex2c.c with
```
mpiexec -n 1 ./ex2c -savef true -output_prefix proffitt -mesh ../share/meshes/proffitt.exo -dt 0.05 -Nt 60000 -sed true -mannings_n 0.06 -ic 2 -h_ic 0.01
``` 

Run ValidateSED.m to show the validation against tRIBS-FEaST

### CODE REVIEW

File: sd/ex2c.c [hunk 1]
```diff
@@ -0,0 +1,3012 @@
...
+
+    if (hr_vec_int[ii] > app->tiny_h) {
+      for (PetscInt jj = 0; jj < ndof-3; jj++) {
+        Cir_vec_int[ii][jj] = fmax(x_ptr[r *ndof + 3 + jj]/hr_vec_int[ii],0.0);
+      }
+    } else {
+      for (PetscInt jj = 0; jj < ndof-3; jj++) {
+        Cir_vec_int[ii][jj] = 0.0;
+      }
+    }
```
Comment (bishtgautam) [inline sd/ex2c.c]: Let's use the same approach for computing `ul_vec_int`/`ur_vec_int` and `Cil_vec_int`/`Cir_vec_int` from the `x_ptr`. So, you could do the following:
-  Move this code in a separate function (e.g. `GetCiFromMass`),
- Get rid of `GetVelocityFromMomentum` and add that code directly here.

File: sd/ex2c.c [hunk 2]
```diff
@@ -0,0 +1,3012 @@
...
+
+    /*
+    for (PetscInt jj = 0; jj < sed->nsed; jj++) {
+      PetscInt index = l * sed->nsed + jj;
+      Cil_vec_int[ii][jj] = sed->Ci[index];
+
+      index = r * sed->nsed + jj;
+      Cir_vec_int[ii][jj] = sed->Ci[index];
+    }
+    */
```
Comment (bishtgautam) [inline sd/ex2c.c]: delete it.

File: sd/ex2c.c [hunk 3]
```diff
@@ -0,0 +1,3012 @@
...
+    hl_vec_bnd[ii]  = x_ptr[l * ndof + 0];
+    hul_vec_bnd[ii] = x_ptr[l * ndof + 1];
+    hvl_vec_bnd[ii] = x_ptr[l * ndof + 2];
+
+    /*
+    for (PetscInt jj = 0; jj < sed->nsed; jj++) {
+      PetscInt index = l * sed->nsed + jj;
+      Cil_vec_bnd[ii][jj] = sed->Ci[index];
+    }
+    */
```
Comment (bishtgautam) [inline sd/ex2c.c]: delete it.

File: sd/ex2c.c [hunk 4]
```diff
@@ -0,0 +1,3012 @@
...
+
+    if (hl_vec_bnd[ii] > app->tiny_h) {
+      for (PetscInt jj = 0; jj < ndof-3; jj++) {
+        Cil_vec_bnd[ii][jj] = fmax(x_ptr[l * ndof + 3 + jj]/hl_vec_bnd[ii],0.0);
+      }
+    } else {
+      for (PetscInt jj = 0; jj < ndof-3; jj++) {
+        Cil_vec_bnd[ii][jj] = 0.0;
+      }
+    }
```
Comment (bishtgautam) [inline sd/ex2c.c]: same comment as the comment for internal edges.

File: sd/ex2c.c [hunk 5]
```diff
@@ -0,0 +1,3012 @@
...
+  /// IC opoition, ic_opt = 0: cold start; ic_opt = 1: dam break condition; ic_opt = 2: constant h
+  PetscInt  ic_opt;
+  /// IC for water depth if ic_opt = 2
+  PetscReal h_ic;
+  /// Flow discharge at outlet
+  PetscReal Qoutlet;
+  /// Flow discharge file
+  FILE      *fq;
+  /// Sediment discharge at outlet
+  PetscReal Soutlet[11];
```
Comment (bishtgautam) [inline sd/ex2c.c]: Let's make it `PetscReal *Soutlet;` and allocate memory based on to `nsed + 1`

File: sd/ex2c.c [hunk 6]
```diff
@@ -0,0 +1,3012 @@
...
+            app->Soutlet[j]         += flux_vec_bnd[ii][j+3] * edgeLen;
+            app->Soutlet[sed->nsed] += flux_vec_bnd[ii][j+3] * edgeLen;
+          }
+        }
+      }
+    }
+  }
+  if ((app->tstep - 1) % qstep == 0) {
+    fprintf(app->fq,"%f\t%f\n",(app->tstep-1)*app->dt,app->Qoutlet);
+    fprintf(app->fs,"%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\n",(app->tstep-1)*app->dt,app->Soutlet[0],
```
Comment (bishtgautam) [inline sd/ex2c.c]: Use a `for` loop to write out `app->Soutlet[...]`

File: sd/ex2c.c [hunk 7]
```diff
@@ -0,0 +1,3012 @@
...
+        }
+        PetscCall(computeSEDMi(app, icell, h));
+        PetscCall(computeSEDohm(ohm,Cd,velocity));
+        sed->Omega[icell] = ohm;
+        PetscCall(computeSEDsource(sed,icell,h,Pr));
+
+        for (PetscInt j = 0; j < sed->nsed; j++) {
+          PetscInt index = icell*nsed+j;
+          Di[j] = (x_ptr[icell * ndof + 3 + j] + dt*f_ptr[icell * ndof + 3 + j] + dt*(sed->Ei[index] + sed->Eri[index] + sed->Ri[index] + sed->Rri[index]))*sed->vset[j]/(h+dt*sed->vset[j]);
+          //Di[j] = sed->vset[j] * sed->Ci[index];
```
Comment (bishtgautam) [inline sd/ex2c.c]: delete it

File: sd/ex2c.c [hunk 8]
```diff
@@ -0,0 +1,3012 @@
...
+          if (icell == 0 && j == 0) {
```
Comment (bishtgautam) [inline sd/ex2c.c:2763]: I believe this was for debugging and can be deleted.

File: sd/ex2c.c [hunk 9]
```diff
@@ -0,0 +1,3012 @@
...
+
+  PetscFunctionReturn(0);
+}
+
+/// Creates a fully initialized RDySed struct from a given DM.
+/// @param [in] dm A DM that provides cell data
+/// @param [out] sed A pointer to an RDySed that stores allocated data.
+///
+/// @return 0 on success, or a non-zero error code on failure
+PetscErrorCode RDySedCreateFromDM(DM dm, RDySed *sed) {
```
Comment (bishtgautam) [inline sd/ex2c.c]: The intent of naming a function as `*FromDM`is to extract some information from the `DM` and save it in a local struct. Here, there isn't any information that is extracted from the `DM` and only `c_end-c_end (=mesh->num_cells)` is used. So, rename the following:

- `RDySedCreateFromDM(...)` to `RDySedCreate(PetscInt ncells, RDySed *sed)`, and
- Rename the existing `RDySedCreate(...)` to `RDySedAllocateMemory(...)`.

**Review by bishtgautam — COMMENTED** [review]
@donghuix Great work getting this implemented that quickly. I have minor comments.

**Comment by bishtgautam** [comment]
You should modify the `.github/workflows/auto_build.yml` to run the newly added example.

File: sd/ex2c.c (hunk 7 above)
Comment (donghuix) [reply sd/ex2c.c]: I am thinking to comment those lines instead of deleting them in case we need to debug. What do you think? @bishtgautam

**Review by bishtgautam — APPROVED** [review]

**Pull request merged**